pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQueryDecorator.java

```diff
@@ -0,0 +1,288 @@
+    }
+    return types;
+  }
+
+  @Override
+  boolean isPreparedFor(int[] paramTypes) {
+    resizeTypes();
+    return isStatementParsed() && super.isPreparedFor(paramTypes);
+  }
+
+  @Override
+  byte[] getEncodedStatementName() {
+    if (batchedEncodedName == null) {
+      String n = super.getStatementName();
+      if (n != null) {
+        batchedEncodedName = Utils.encodeUTF8(n);
+      }
+    }
+    return batchedEncodedName;
+  }
+
+  @Override
+  void setStatementName(String statementName) {
+    if (statementName == null) {
+      batchedEncodedName = null;
+      super.setStatementName(null);
+    } else {
+      super.setStatementName(statementName);
+      batchedEncodedName = Utils.encodeUTF8(statementName);
+    }
+  }
+
+  /**
+   * Detect when the vanilla prepared type meta data is out of date.
+   *
+   * @param preparedTypes
+   *          meta data to compare with
+   * @return boolean value indicating if internal type information needs
+   *         updating
+   */
+  private boolean isOriginalStale(int[] preparedTypes) {
+    if (isPreparedTypesSet) {
+      return false;
+    }
+    if (preparedTypes == null) {
+      return false;
+    }
+    if (preparedTypes.length == 0) {
+      return false;
+    }
+    if (preparedTypes.length < originalPreparedTypes.length) {
+      return false;
+    }
+    int maxPos = originalPreparedTypes.length - 1;
+    for (int pos = 0; pos <= maxPos; pos += 1) {
+      if (originalPreparedTypes[pos] == UNSPECIFIED
+          && preparedTypes[pos] != UNSPECIFIED) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void updateOriginal(int[] preparedTypes) {
+    if (preparedTypes == null) {
+      return;
+    }
+    if (preparedTypes.length == 0) {
+      return;
+    }
+    isPreparedTypesSet = true;
+    int maxPos = originalPreparedTypes.length - 1;
+    for (int pos = 0; pos <= maxPos; pos++) {
+      if (preparedTypes[pos] != UNSPECIFIED) {
+        if (originalPreparedTypes[pos] == UNSPECIFIED) {
+          originalPreparedTypes[pos] = preparedTypes[pos];
+        }
+      } else {
+        isPreparedTypesSet = false;
+      }
+    }
+  }
+
+  private boolean isStatementParsed() {
+    return isParsed.equals(getBindPositions());
+  }
+
+  /**
+   * Method to receive notification of the parsed/prepared status of the
+   * statement.
+   *
+   * @param prepared
+   *          state
+   */
+  public void registerQueryParsedStatus(boolean prepared) {
+    isParsed = getBindPositions();
+  }
+
+  @Override
+  String getNativeSql() {
+    // dynamically build sql with parameters for each batch
+    if (super.getNativeSql() == null) {
+      return "";
+    }
+    int c = super.getNativeQuery().bindPositions.length;
+    int bs = getBatchSize();
+    calculateLength(super.getNativeSql().length(), c, bs - 1 );
+    StringBuilder s = new StringBuilder(length).append(super.getNativeSql());
+    for (int i = 2; i <= bs; i += 1) {
```

Comment:
This looks sufficiently complex to warrant a comment, esp. the pre-increment bit